oepcie/yuanphil_impl/oe_ctx_impl.c

```diff
@@ -0,0 +1,221 @@
+#include "oe_ctx_impl.h"
+
+void _oe_create(oe_ctx_impl_t* ctx){
+	ctx -> header.path = malloc(sizeof(char));
+	ctx -> config.path = malloc(sizeof(char));
+	ctx -> data.path = malloc(sizeof(char));
+
+	ctx -> initialized = 'F';
+	ctx -> config_id = -1;
+}
+
+int _oe_init(oe_ctx_impl_t* ctx){
+	ctx -> initialized = 'T';
+
+	// Successfully open all filestreams
+	ctx -> header.fid = open(ctx -> header.path, O_RDONLY);
+	if(ctx -> header.fid == -1){
+		printf("Failed to open header filestream, returning with error\n");
+		return -1;
+	}
+	ctx -> config.fid = open(ctx -> config.path, O_RDWR);
+	if(ctx -> config.fid == -1){
+		printf("Failed to open config filestream, returning with error\n");
+		return -1;
+	}
+	ctx -> data.fid = open(ctx -> data.path, O_RDONLY);
+	if(ctx -> data.fid == -1){
+		printf("Failed to open data filestream, returning with error\n");
+		return -1;
+	}
+
+	read(ctx -> header.fid, &(ctx -> config_id), sizeof(int));
+
+	int device_id = 0;
+	int num_devices = 0;
+
+	while(device_id != -1 && ++num_devices){
+		if(device_id > MAX_DEVICE_ID || device_id < 0){
+			printf("Invalid device detected, returning error\n");
+			return -1;
+		}
+		read(ctx -> header.fid, &device_id, sizeof(int));
+		printf("Read device_id %d\n", device_id); // debug
+	}
+	printf("Devices detected: %d\n", num_devices);
+
+	ctx -> map.num_dev = num_devices;
+	ctx -> map.devs = calloc(num_devices, sizeof(device_t));
+
+	// return to start of device_ids
+	lseek(ctx -> header.fid, 4, SEEK_SET);
+
+	for(int i = 0; i < num_devices; i++){
+		int id = read(ctx -> header.fid, &id, sizeof(int));
+		ctx -> map.devs[i] = DEVICES[id];
+	}
+
+	// calculate read offsets
+	for(int i = 1; i < num_devices; i++){
+		ctx -> map.devs[i].read_offset = ctx -> map.devs[i - 1].read_offset +
+			ctx -> map.devs[i - 1].read_size;
+	}
+
+	/*
+	 * Not implemented - Write Offsets... Aren't exactly sure what/if these are
+	 * used yet.
+	 */
+
+	 return 1;
+
+}
+
+int _oe_set_opt(oe_ctx_impl_t* ctx, int option, const char* value){
+	if(ctx -> initialized == 'T'){
+		printf("Already initialized, no more changing options\n");
+		return -1;
+	}
+
+	switch (option){
+		case OE_HEADER_STREAMPATH:
+			ctx -> header.path = realloc(ctx -> header.path, strlen(value) + 1);
```

Comment:
check for ctx->header.path == NULL return value from realloc and return error if so (http://en.cppreference.com/w/c/memory/realloc)